daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala

```diff
@@ -46,6 +46,42 @@ sealed abstract class Value[+Cid] extends Product with Serializable {
       case ValueMap(x) => ValueMap(x.mapValue(_.mapContractId(f)))
     }
 
+  // TODO (Nick Smith) - Still not tail-recursive
+  def mapContractIdOpt[Cid2](f: Cid => Option[Cid2]): Option[Value[Cid2]] = {
+    @tailrec def _flatten[A](
+        in: ImmArray[Option[A]],
+        out: ImmArray.ImmArraySeq[A]): Option[ImmArray[A]] =
+      in match {
+        case ImmArray() => Some(out.toImmArray)
+        case ImmArrayCons(None, _) => None
+        case ImmArrayCons(Some(v), rest) => _flatten(rest, out :+ v)
+      }
+    def flatten[A](in: ImmArray[Option[A]]): Option[ImmArray[A]] =
+      _flatten(in, ImmArray.ImmArraySeq.empty)
+
+    this match {
+      case prim @ (ValueBool(_) | ValueDate(_) | ValueDecimal(_) | ValueInt64(_) | ValueParty(_) |
+          ValueText(_) | ValueTimestamp(_) | ValueUnit) =>
+        Some(prim.asInstanceOf[Value[Cid2]])
```

Comment:
Please expand all these cases instead of using `asInstanceOf`. You can also follow the pattern of `GenNode.LeafOnlyNode` and the way some of the `GenNode` constructors inherit from it instead of `GenNode` directly to introduce, say, a `ValueCidless` for these cases as GADT evidence that you can get away with this, then use that as this case.